base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,813 @@
+
+        for (int pos = 1; pos > 0 && !Info.Error;)
+        {
+            pos = files.Find(L'|');
+            CStringW item = pos <= 0 ? files : files.Left(pos);
+            files = files.Mid(pos + 1);
+            InstallFiles(tempdir, item, installdir);
+        }
+
+        if (!Info.MainApp.IsEmpty())
+        {
+            AddUninstallOperationsFromDB(DB_GENINST_DELREG, UNOP_REGKEY);
+            AddUninstallOperationsFromDB(DB_GENINST_DELREGEMPTY, UNOP_EMPTYREGKEY);
+            AddUninstallOperationsFromDB(DB_GENINST_DELFILE, UNOP_FILE, CStringW(L"\\"));
+            AddUninstallOperationsFromDB(DB_GENINST_DELDIR, UNOP_DIR, CStringW(L"\\"));
+            AddUninstallOperationsFromDB(DB_GENINST_DELDIREMPTY, UNOP_EMPTYDIR, CStringW(L"\\"));
+
+            AddEntry(UNOP_EMPTYDIR, L"\\");
+            WriteArpEntry(L"DisplayName", AppName);
+            WriteArpEntry(L"InstallLocation", Info.InstallDir); // Note: This value is used by the uninstaller!
+
+            LPWSTR p = tmp.GetBuffer(1 + MAX_PATH);
+            p[0] = L'\"';
+            GetModuleFileName(NULL, p + 1, MAX_PATH);
+            tmp.ReleaseBuffer();
+            UINT cch = tmp.GetLength(), bitness = IsSystem64Bit() ? 64 : 32;
+            WCHAR modechar = Info.PerUser ? 'U' : 'M';
+            LPCWSTR unparamsfmt = L"\" /" CMD_KEY_UNINSTALL L" /K%s \"%c%d\\%s\"";
+            (tmp = tmp.Mid(0, cch)).AppendFormat(unparamsfmt, L"", modechar, bitness, arpkeyname);
+            WriteArpEntry(L"UninstallString", tmp);
+            (tmp = tmp.Mid(0, cch)).AppendFormat(unparamsfmt, L" /S", modechar, bitness, arpkeyname);
+            WriteArpEntry(L"QuietUninstallString", tmp);
+
+            if (GetCustomIconPath(Info, tmp) != S_OK)
+                tmp = Info.MainApp;
+            WriteArpEntry(L"DisplayIcon", tmp);
+
+            if (*GetCommonString(DB_VERSION, tmp))
+                WriteArpEntry(L"DisplayVersion", tmp);
+
+            SYSTEMTIME st;
+            GetSystemTime(&st);
+            tmp.Format(L"%.4u%.2u%.2u", st.wYear, st.wMonth, st.wDay);
+            WriteArpEntry(L"InstallDate", tmp);
+
+            if (*GetCommonString(DB_PUBLISHER, tmp))
+                WriteArpEntry(L"Publisher", tmp);
+
+#if DBG
+            tmp.Format(L"sys64=%d", IsSystem64Bit());
+            WriteArpEntry(L"_DEBUG", tmp);
+#endif
+        }
+
+        if (!Info.Error && Info.ShortcutFile)
+        {
+            CreateShortcut(Info.MainApp);
+        }
+    }
+
+    DeleteDirectoryTree(tempdir.GetString());
+    RemoveDirectory(installdir.GetString()); // This is harmless even if we installed something
+    SendMessage(Info.hDlg, IM_PROGRESS, 0, 0); // Cleanup
+    return 0;
+}
+
+static DWORD CALLBACK
+UIWorkerThread(LPVOID Parameter)
+{
+    ((LPTHREAD_START_ROUTINE)Parameter)(NULL);
+    return SendMessage(g_pInfo->hDlg, IM_END, 0, 0);
+}
+
+static INT_PTR CALLBACK
+UIDlgProc(HWND hDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
```

Comment:
The rest of the code uses ATL for the UI, any reason why this code doesn't?

Reply:
It's just babysitting the progress bar, no need to complicate things.